src/analysis/inner_product_space/l2_space.lean

```diff
@@ -95,4 +142,284 @@ lemma inner_eq_tsum (f g : lp G 2) : ⟪f, g⟫ = ∑' i, ⟪f i, g i⟫ := rfl
+    intros s,
+    refine sum_mem (supr (λ i, (V i).to_linear_map.range)) _,
+    intros i hi,
+    refine mem_supr_of_mem i _,
+    exact linear_map.mem_range_self _ (f i) },
+  { apply topological_closure_minimal,
+    { refine supr_le _,
+      rintros i x ⟨x, rfl⟩,
+      use lp.single 2 i x,
+      convert hV.linear_isometry_apply_single _ },
+    exact hV.linear_isometry.isometry.uniform_inducing.is_complete_range.is_closed }
+end
+
+/-- A mutually orthogonal family of complete subspaces of `E`, whose range is dense in `E`, induces
+a linear isometry from E to `lp 2` of the subspaces.
+
+Note that this goes in the opposite direction from `orthogonal_family.linear_isometry`. -/
+noncomputable def linear_isometry_equiv [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) :
+  E ≃ₗᵢ[𝕜] lp G 2 :=
+linear_isometry_equiv.symm $
+linear_isometry_equiv.of_surjective
+hV.linear_isometry
+begin
+  refine linear_map.range_eq_top.mp _,
+  rw ← hV',
+  rw hV.range_linear_isometry,
+end
+
+protected lemma linear_isometry_equiv_symm_apply [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) (w : lp G 2) :
+  (hV.linear_isometry_equiv hV').symm w = ∑' i, V i (w i) :=
+by simp [orthogonal_family.linear_isometry_equiv, orthogonal_family.linear_isometry_apply]
+
+protected lemma has_sum_linear_isometry_equiv_symm [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) (w : lp G 2) :
+  has_sum (λ i, V i (w i)) ((hV.linear_isometry_equiv hV').symm w) :=
+by simp [orthogonal_family.linear_isometry_equiv, orthogonal_family.has_sum_linear_isometry]
+
+@[simp] protected lemma linear_isometry_equiv_symm_apply_single [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) {i : ι} (x : G i) :
+  (hV.linear_isometry_equiv hV').symm (lp.single 2 i x) = V i x :=
+by simp [orthogonal_family.linear_isometry_equiv, orthogonal_family.linear_isometry_apply_single]
+
+@[simp] protected lemma linear_isometry_equiv_symm_apply_dfinsupp_sum_single
+  [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) (W₀ : Π₀ (i : ι), G i) :
+  (hV.linear_isometry_equiv hV').symm (W₀.sum (lp.single 2)) = (W₀.sum (λ i, V i)) :=
+by simp [orthogonal_family.linear_isometry_equiv,
+  orthogonal_family.linear_isometry_apply_dfinsupp_sum_single]
+
+@[simp] protected lemma linear_isometry_equiv_apply_dfinsupp_sum_single
+  [Π i, complete_space (G i)]
+  (hV' : (⨆ i, (V i).to_linear_map.range).topological_closure = ⊤) (W₀ : Π₀ (i : ι), G i) :
+  (hV.linear_isometry_equiv hV' (W₀.sum (λ i, V i)) : Π i, G i) = W₀ :=
+begin
+  rw ← hV.linear_isometry_equiv_symm_apply_dfinsupp_sum_single hV',
+  rw linear_isometry_equiv.apply_symm_apply,
+  ext i,
+  simp [dfinsupp.sum, lp.single_apply] {contextual := tt},
+end
```

Comment:
can you add docstrings for these lemmas? Parsing them is not that easy.

Reply:
Done!